smtpserver/server.go

```diff
@@ -1752,6 +1803,26 @@ func (c *conn) cmdData(p *parser) {
 		}
 	}
 
+	// Now that we have all the whole message (envelope + data), we can check if the SMTPUTF8 extension is required.
+	// The check happens only when the client required the SMTPUTF8 extension.
+	var part *message.Part
+	if c.smtputf8 {
+		// Try to parse the message.
+		// Do nothing if something bad happen during Parse and Walk, just keep the current value for c.msgsmtputf8.
+		p, err := message.Parse(c.log.Logger, true, dataFile)
+		if err == nil {
+			// Message parsed without error. Keep the result to avoid parsing the message again.
+			part = &p
+			err = part.Walk(c.log.Logger, nil)
+			if err == nil {
+				c.msgsmtputf8 = c.isSmtpUtf8Required(part)
+			}
+		}
+		if c.smtputf8 != c.msgsmtputf8 {
+			c.log.Debug("SMTPUTF8 flag changed", slog.Bool("received SMTPUTF8", c.smtputf8), slog.Bool("evaluated SMTPUTF8", c.msgsmtputf8))
```

Comment:
style nit: the attributes shouldn't have spaces. it may even break the logfmt we are generating. we could just keep the names "smtputf8" and "msgsmtputf8". i also try to keep the messages lowercase, though that's not applied consistently.